### ISSUE TITLE
Display labels and ticks on a clipped (custom shape) map layout. 

### ISSUE THREAD

**Issue by MarshEnterprises** [qgis/QGIS#47309]
### Feature description

It would be great if the map item in the print layout could be a shape other than rectangular. Clipping the map is currently supported, but as far as I can tell it is not possible to add ticks and x/y labels to a clipped map (see below).

![example](https://user-images.githubusercontent.com/52695265/153652282-d19091ff-ce34-4580-91d5-92dee9b6357b.png)

This has already been suggested here: https://github.com/qgis/QGIS/pull/38024

It would be particularly useful for publications where linear features such as rivers or the coastline are of interest and space is an issue. 

Thanks for all the hard work! QGIS keeps on getting better.

**Comment by antoniolocandro** [comment]
I agree this would be a nice feature to be added, I suppose the grid should be clipped at the same time the map is clipped in order to achieve the result 

In this sample the map is clipped to the outline of the magenta line but the grid extends to the original extent of the map frame instead of the clipped feature, in theory if the grid was to be clipped also that should allow for the placement of the labels. Do not know how hatching may work though

![image](https://user-images.githubusercontent.com/4627073/155385779-bd05cc40-8867-42e0-a73f-a7487a6b7dc3.png)
